libraries/mbed/targets/hal/TARGET_NXP/TARGET_LPC81X/i2c_api.c

```diff
@@ -145,62 +149,82 @@ void i2c_frequency(i2c_t *obj, int hz) {
 // because something is setup wrong (e.g. wiring), and we don't need to programatically
 // check for that
 
+//New version WH, Tested OK for Start and Repeated Start
+//Old version was Wrong: Calls i2c_start without setting address, i2c_do_read continues before checking status, status check for wrong value
 int i2c_read(i2c_t *obj, int address, char *data, int length, int stop) {
     int count, status;
     
-    i2c_start(obj);
-    
-    status = i2c_do_write(obj, (address | 0x01), 1);
-    if (status != 0x01) {
+    //Store the address+RD and then generate STA
+    I2C_DAT(obj) = address | 0x01;
+    i2c_start(obj);    
+
```

Comment:
Hi Martin, I think I explained this in private email to you regarding the
I2C problems on the lpc812. 

Anyhow, the issue here is that the current implementation of i2c_start(obj)
on lpc812 is wrong. The function generates a Start condition, however that
is BEFORE the address is stored in the DAT register in the call to
i2c_do_write(obj, address | 0x01, 1). The lpc812 I2C engine is very
different from the lpc1768 and a call to i2c_start() will generate a start
AND immediately continue with sending the content of the DAT register. The
result is unpredictable. A random address may be transmitted when there is
delay between calling start and loading the DAT register. My patch in the
block operations fixes that. However, the problem remains when a program
directly calls the i2c.start() and byte i2c.write(slaveaddress) methods. Any
delay (eg interrupt) between them will lead to a crash. My proposal is to
use a new i2c.start() that takes the address as additional parameter.

 

Wim

 

  _____  

From: mbed-devel@googlegroups.com [mailto:mbed-devel@googlegroups.com] On
Behalf Of Martin Kojtal
Sent: maandag 26 januari 2015 7:44
To: mbedmicro/mbed
Subject: Re: [mbed] I2C blockread and blockwrite fixed for LPC812 (#862)

 

In libraries/mbed/targets/hal/TARGET_NXP/TARGET_LPC81X/i2c_api.c
<https://github.com/mbedmicro/mbed/pull/862#discussion_r23514306> :

>  int i2c_read(i2c_t *obj, int address, char *data, int length, int stop) {
>      int count, status;
>      
> -    i2c_start(obj);
> -    
> -    status = i2c_do_write(obj, (address | 0x01), 1);
> -    if (status != 0x01) {
> +    //Store the address+RD and then generate STA
> +    I2C_DAT(obj) = address | 0x01;
> +    i2c_start(obj);    
> +

Why this was changed from start, address ? 

-
Reply to this email directly or view it on
<https://github.com/mbedmicro/mbed/pull/862/files#r23514306>  GitHub.
<https://github.com/notifications/beacon/AFAC17XU1h6NN3DxoBcmcSSGIc6I1nttks5
nldmSgaJpZM4DW8fS.gif> 


-- 
You received this message because you are subscribed to the Google Groups
"mbed-devel" group.
To unsubscribe from this group and stop receiving emails from it, send an
email to mbed-devel+unsubscribe@googlegroups.com.
For more options, visit https://groups.google.com/d/optout.